envelope.go

```diff
@@ -67,14 +67,27 @@ func (w *envelopeWriter) Marshal(message any) *Error {
 		}
 		return nil
 	}
-	raw, err := w.codec.Marshal(message)
+
+	buffer := w.bufferPool.Get()
+	defer w.bufferPool.Put(buffer)
+
+	// Reuse byte buffer if codec supports MarshalAppend.
+	var (
+		raw []byte
+		err error
+	)
+	if c, ok := w.codec.(marshalAppend); ok {
+		raw, err = c.MarshalAppend(buffer.Bytes(), message)
+	} else {
+		raw, err = w.codec.Marshal(message)
+		if len(raw) > 0 {
```

Comment:
nit: Would prefer we only use the value if the error was non-nil - some static analysis tools ensure this.

```suggestion
		// We can't avoid allocating the byte slice, so we may as well reuse it once
		// we're done with it.
		if err == nil && len(raw) > 0 {
```